### PR TITLE
feat(feg): support Gx/Gy interfaces request timeout configuration from mconfig

### LINKED ISSUE #12247: session_proxy: configurable RequestTimeout
When sending a diameter message we can configure the diameter client to timeout after N seconds.

After those N seconds, gx and gy clients will time out and forget about that message (see that action on both clients)
[gx](http://github.com/magma/magma/blob/7c9fa67b8881b6d4653186c8f92e93e0a28d4601/feg/gateway/services/session_proxy/servicers/policy.go#L142-L142)
[gy](http://github.com/magma/magma/blob/7c9fa67b8881b6d4653186c8f92e93e0a28d4601/feg/gateway/services/session_proxy/servicers/credits.go#L202-L202)

Session_proxy has this timeout hardcoded [here](http://github.com/magma/magma/blob/7c9fa67b8881b6d4653186c8f92e93e0a28d4601/feg/gateway/services/session_proxy/session_proxy/main.go#L129-L129) to 3 seconds. That means that if PCRF or OCS takes more than 3 seconds to answer, FEG will not be able to handle that message and will trigger error `no Key/channel registered for message:`

In order to fix that we can do that timeout to be configurable adding it to the swagger config and configuring it [here](http://github.com/magma/magma/blob/7c9fa67b8881b6d4653186c8f92e93e0a28d4601/feg/gateway/services/session_proxy/session_proxy/main.go#L129-L129)

### PR DESCRIPTION
**Pull request by malcolmmadsheep**
Signed-off-by: Vladyslav Zhuk <vzhuk@freedomfi.com>

## Summary

1. Provide support of configuration of request timeout values for Gx and Gy client interfaces if feg/session_proxy component.
1. Add support for request timeout to s6a_proxy client config.
1. Fix `s8_cli` tests in feg

Fixes https://github.com/magma/magma/issues/12247

## Test Plan

Add unit tests for Gx and Gy client configs generation from file.

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: feg/gateway/services/session_proxy/credit_control/gy/config_test.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+					"servers": [{
+						"protocol": "sctp",
+						"address": "1.1.1.1:9999",
+						"retransmits": 5,
+						"watchdog_interval": 1,
+						"retry_count": 5,
+						"product_name": "magma_test",
+						"realm": "local.openair4G.eur",
+						"host": "local.magma-oai.openair4G.eur",
+						"dest_host":"magma-oai.openair4G.eur",
```
Comment (maxhbr) [inline feg/gateway/services/session_proxy/credit_control/gy/config_test.go]:
```suggestion
						"dest_host": "magma-oai.openair4G.eur",
```

File: feg/gateway/services/session_proxy/credit_control/gy/config_test.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+						"dest_realm":"openair4G.eur",
```
Comment (maxhbr) [inline feg/gateway/services/session_proxy/credit_control/gy/config_test.go]:
```suggestion
						"dest_realm": "openair4G.eur",
```

File: feg/gateway/services/session_proxy/credit_control/gx/config_test.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+					"servers": [{
+						"protocol": "sctp",
+						"address": "1.1.1.1:9999",
+						"retransmits": 5,
+						"watchdog_interval": 1,
+						"retry_count": 5,
+						"product_name": "magma_test",
+						"realm": "local.openair4G.eur",
+						"host": "local.magma-oai.openair4G.eur",
+						"dest_host":"magma-oai.openair4G.eur",
```
Comment (maxhbr) [inline feg/gateway/services/session_proxy/credit_control/gx/config_test.go]:
```suggestion
						"dest_host": "magma-oai.openair4G.eur",
```

File: feg/gateway/services/session_proxy/credit_control/gx/config_test.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+						"dest_realm":"openair4G.eur",
```
Comment (maxhbr) [inline feg/gateway/services/session_proxy/credit_control/gx/config_test.go]:
```suggestion
						"dest_realm": "openair4G.eur",
```

File: feg/gateway/diameter/config.go [hunk 5]
```diff
@@ -133,6 +135,9 @@ func (srcCfg *DiameterClientConfig) FillInDefaults() *DiameterClientConfig {
 	if cfg.Retransmits == 0 {
 		cfg.Retransmits = 3
 	}
+	if cfg.RequestTimeout == 0 {
+		cfg.RequestTimeout = 3
```
Comment (maxhbr) [inline feg/gateway/diameter/config.go]: why does that not use `DefaultRequestTimeoutSeconds`?

**Review by maxhbr — APPROVED** [review]
LGTM, up to some minor comments

File: vscode-workspaces/multiroot-go.code-workspace [hunk 6]
```diff
@@ -41,9 +41,9 @@
 			"name": "src"
 		},
 		{
-			"path": "../wifi/cloud/go",
-			"name": "wifi/cloud",
-		},
+			"path": "../feg/gateway",
+			"name": "feg/gateway"
+		}
```
Comment (uri200) [inline vscode-workspaces/multiroot-go.code-workspace:46]: why has this changed

Comment (maxhbr) [reply vscode-workspaces/multiroot-go.code-workspace:46]: `wifi/cloud` is no longer in the source tree. `feg/gateway` was missing. Looks valid to me

**Review by uri200 — COMMENTED** [review]
it may be good to look at `s6a_proxy` too. It has the same problem as session proxy

File: vscode-workspaces/multiroot-go.code-workspace (hunk 6 above)
Comment (malcolmmadsheep) [reply vscode-workspaces/multiroot-go.code-workspace:46]: when I'm trying to open this workspace file I see message that there is no `wifi/cloud` destination, and in `feg/gateway` there is go code that we're working with

**Comment by uri200** [comment]
to make sure format is proper, just run 
```
cd magma/feg/gateway/docker
./build.py -c
```
Thats how you run the pre-commit tests. That will run the linter. vet, and tests. 

You can also achieve the same doing
```
./build.py -m
make precommit 
(inside the docker container)
```

**Comment by malcolmmadsheep** [comment]
Hi, @uri200 thanks! I will take a look today on `s6a_proxy` as well and submit changes with precommit checks. I see some github actions have been failed. Should all of them to be green to merge PR?

**Comment by malcolmmadsheep** [comment]
@uri200 I've added support for `requestTimeout` for client config in `s6a_proxy` service, but it's not used anywhere, because in case of Gy/Gx we have session manager which is responsible for waiting, but for `s6a_pxoxy` I couldn't find any usage of `requestTimeout`.

When I tried:
```bash
./build.py -m
make precommit 
(inside the docker container)
```

It fails with the `<service_name>.depends_on contains an invalid type, it should be an array` - seems like it was a feature of docker-compose v2, and it was removed in v3. To make `./build.py` work I just transform `depends_on` property into array. I've opened an issue regarding this https://github.com/magma/magma/issues/12356.

**Comment by uri200** [comment]
That is right, it looks like s6a_proxy just uses send message, no time out at all. Maybe it can be added in the future


http://github.com/magma/magma/blob/master/feg/gateway/services/s6a_proxy/servicers/ai.go#L69-L69

File: feg/gateway/services/session_proxy/credit_control/gx/config_test.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+	assert.NoError(t, err)
+	cliConfig := GetGxClientConfiguration()[0]
+
+	assert.Equal(t, "local.magma-oai.openair4G.eur", cliConfig.Host)
+	assert.Equal(t, "local.openair4G.eur", cliConfig.Realm)
+	assert.Equal(t, "magma_test", cliConfig.ProductName)
+	assert.Equal(t, uint(1), cliConfig.WatchdogInterval)
+	assert.Equal(t, uint(5), cliConfig.RetryCount)
+	assert.Equal(t, uint(5), cliConfig.Retransmits)
+	assert.Equal(t, uint(10), cliConfig.RequestTimeout)
```
Comment (uri200) [inline feg/gateway/services/session_proxy/credit_control/gx/config_test.go:66]: Can you make sure you cover all the fields 
```
			{
				Host:               diameter.GetValueOrEnv(diameter.HostFlag, GxDiamHostEnv, diameter.DiamHost),
				Realm:              diameter.GetValueOrEnv(diameter.RealmFlag, GxDiamRealmEnv, diameter.DiamRealm),
				ProductName:        diameter.GetValueOrEnv(diameter.ProductFlag, GxDiamProductEnv, diameter.DiamProductName),
				AppID:              diam.GX_CHARGING_CONTROL_APP_ID,
				WatchdogInterval:   diameter.DefaultWatchdogIntervalSeconds,
				RequestTimeout:     diameter.DefaultRequestTimeoutSeconds,
				RetryCount:         uint(retries),
				Retransmits:        uint(retransmits),
				SupportedVendorIDs: diameter.GetValueOrEnv("", GxSupportedVendorIDsEnv, ""),
			},
		}
		```

File: feg/gateway/services/session_proxy/credit_control/gy/config_test.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+	assert.NoError(t, err)
+	cliConfig := GetGyClientConfiguration()[0]
+
+	assert.Equal(t, "local.magma-oai.openair4G.eur", cliConfig.Host)
+	assert.Equal(t, "local.openair4G.eur", cliConfig.Realm)
+	assert.Equal(t, "magma_test", cliConfig.ProductName)
+	assert.Equal(t, uint(1), cliConfig.WatchdogInterval)
+	assert.Equal(t, uint(5), cliConfig.RetryCount)
+	assert.Equal(t, uint(5), cliConfig.Retransmits)
+	assert.Equal(t, uint(10), cliConfig.RequestTimeout)
```
Comment (uri200) [inline feg/gateway/services/session_proxy/credit_control/gy/config_test.go:70]: same here

File: cwf/cloud/go/go.sum [hunk 9]
```diff
@@ -92,6 +92,7 @@ github.com/dgrijalva/jwt-go v3.2.0+incompatible h1:7qlOGliEKZXTDg6OTjfoBKDXWrumC
 github.com/dgrijalva/jwt-go v3.2.0+incompatible/go.mod h1:E3ru+11k8xSBh+hMPgOLZmtrrCbhqsmaPHjLKYnJCaQ=
 github.com/dgryski/go-farm v0.0.0-20190423205320-6a90982ecee2/go.mod h1:SqUrOPUnsFjfmXRMNPybcSiG0BgUW2AuFH8PAnS2iTw=
 github.com/dgryski/go-sip13 v0.0.0-20181026042036-e10d5fee7954/go.mod h1:vAd38F8PWV+bWy6jNmig1y/TA+kYO4g3RSRF0IAv0no=
+github.com/dnephin/pflag v1.0.7/go.mod h1:uxE91IoWURlOiTUIA8Mq5ZZkAv3dPUfZNaT80Zm7OQE=
```
Comment (uri200) [inline cwf/cloud/go/go.sum]: I would remove all the go.sum files from this PR

Sometimes they get autogenerated by the IDE

**Comment by uri200** [comment]
yes you need to remove the go.sum , otherwise `insyc checking job` will not work


```
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   ../../../cwf/cloud/go/go.sum
	modified:   ../../../dp/cloud/go/go.sum
	modified:   ../../../feg/cloud/go/go.sum
	modified:   ../../../lte/cloud/go/go.sum
	modified:   ../go/go.mod
	modified:   ../go/go.sum
```

File: feg/gateway/services/session_proxy/credit_control/gx/config_test.go (hunk 7 above)
Comment (malcolmmadsheep) [reply feg/gateway/services/session_proxy/credit_control/gx/config_test.go:66]: I've added also test for values from env variables. Before read config I set env value and unset it right after test finished. I also thought about saving original value of env var, and set it back after test. What do you think is the most preferable way?

Comment (uri200) [reply feg/gateway/services/session_proxy/credit_control/gx/config_test.go:66]: This looks fine. I was more concerned about covering all the fields

If you want to test the other path, you can use `os.Setenv`

File: feg/gateway/tools/s8_cli/main_test.go [hunk 10]
```diff
@@ -13,5 +13,8 @@ func TestS8Cli(t *testing.T) {
 	// replace the arguments
 	os.Args = []string{"s8_cli", "cs", "-test", "-delete", "0"}
 	fmt.Printf("\ncommand to run: %+v\n", os.Args)
-	main()
+	exitCode := run()
+	if exitCode != 0 {
+		t.Fatalf("expected exit code to be 0, got: %d", exitCode)
+	}
```
Comment (uri200) [inline feg/gateway/tools/s8_cli/main_test.go]: I don't remember now but I think there was a reason why we were just exiting. Maybe there are some notes on documentation.  Were you able to run the client and test the responses out?

In any case, this is out of scope of this PR. If this needs to be changed lets create a different PR

Comment (malcolmmadsheep) [reply feg/gateway/tools/s8_cli/main_test.go]: Ok, sure, I agree that if it was done so intentionally, than this fix should be done in different PR. `make test` just fails when I'm running it in `feg/gateway`, so I thought that should help

**Comment by malcolmmadsheep** [comment]
removed commit with `s8_cli` tests fix and merged after review commits into one

**Review by uri200 — APPROVED** [review]
